### PR TITLE
[Sampling] NodeDataLoader for node classification

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This PR
* Encapsulates neighborhood samplers into `BlockSampler` objects.  Simple multi-layer neighbor sampling is implemented as `MultiLayerNeighborSampler` class which is a subclass of `BlockSampler`.
* Implements `NodeCollator` which is dedicated for iterating over a set of nodes, constructing list of blocks for GNN computation by the specified `BlockSampler`.

## Example usage

```python
# Create a 3-layer neighborhood sampler that samples 5, 10, and 15 neighbors
# on the first, second, and third GNN layer.
sampler = dgl.sampling.MultiLayerNeighborSampler([5, 10, 15])
# Create a collator that provides an iterable over the dictionary of node IDs, as well as a collator
# function that generates computation dependency graphs.
collator = dgl.sampling.NodeCollator(g, train_nid, sampler)
# Create a PyTorch DataLoader
dataloader = torch.utils.data.DataLoader(
    collator.dataset, collate_fn=collator.collate, batch_size=1024, num_workers=4)

for blocks in dataloader:
    input_nodes = blocks[0].srcdata[dgl.NID]
    output_nodes = blocks[-1].dstdata[dgl.NID]
    # ...
```

## TODOs
~~(I may need help on MXNet and Tensorflow DataLoader solutions):~~  Not anymore since `Collator` object is no longer a `DataLoader`.
- [x] Switch to `NodeDataLoader` for all node classification examples.
  - Currently GraphSAGE, RGCN, and OGB GraphSAGE submission are switched.
- [x] ~~Implement MXNet DataLoader (@aksnzhy @zheng-da I remember Gluon has a similar DataLoader interface?  Is it a drop-in replacement of PyTorch's DataLoader?)~~ No longer necessary.
- [x] ~~Implement Tensorflow DataLoader (@VoVAllen )~~ No longer necessary.
- [x] Check if model performance is affected.

## Other fixes
* Fixes RGCN minibatch training crashing with self loop on.
* Fixes `DGLHeteroGraph.find_edges` crashing with empty edge ID tensor.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by classicsong** [comment]
The plan is to build dgl DataLoader instead of using PyTorch DataLoader?

**Comment by BarclayII** [comment]
> The plan is to build dgl DataLoader instead of using PyTorch DataLoader?

Yes.  Essentially it's an adapter to PyTorch/MXNet/Tensorflow(?) DataLoaders.  I found it not that straightforward to write a native PyTorch DataLoader for training on heterogeneous graphs, where the node IDs will be dictionary of node types and tensors.

**Comment by VoVAllen** [comment]
How to deal with multiprocessing with TF/MXNet?

**Comment by BarclayII** [comment]
> How to deal with multiprocessing with TF/MXNet?

For MXNet maybe we can use Gluon's DataLoader?

For TF I don't know.  Didn't find the equivalent of `collate_fn`.

File: docs/source/api/python/sampling.rst [hunk 1]
```diff
@@ -30,3 +30,15 @@ Builtin sampler classes for more complicated sampling algorithms
...
 .. autoclass:: RandomWalkNeighborSampler
 .. autoclass:: PinSAGESampler
+
+Neighborhood samplers for multilayer GNNs
+-----------------------------------------
+.. autoclass:: MultiLayerNeighborSampler
+
+Data loaders for minibatch iteration
+------------------------------------
+.. autoclass:: NodeDataLoader
```
Comment (classicsong) [inline docs/source/api/python/sampling.rst:40]: We have NodeCollator instead of NodeDataLoader, right?

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py [hunk 2]
```diff
@@ -150,10 +118,11 @@ def main(args):
         if epoch > 3:
             t0 = time.time()
 
-        for i, (seeds, blocks) in enumerate(loader):
+        for i, blocks in enumerate(loader):
+            seeds = blocks[-1].dstnodes[category].data[dgl.NID]
```
Comment (classicsong) [inline examples/pytorch/rgcn-hetero/entity_classify_mb.py:122]: Can we return the seeds from the loader iter?

File: python/dgl/nn/mxnet/conv/graphconv.py [hunk 3]
```diff
@@ -109,8 +110,14 @@ def forward(self, graph, feat, weight=None):
...
-        feat : mxnet.NDArray
-            The input feature.
+        feat : mxnet.NDArray or pair of mxnet.NDArray
+            If a single tensor is given, represents the input feature of shape :math:`(N, D_{in})`
+            where :math:`D_{in}` is size of input feature, :math:`N` is the number of nodes.
+            If a pair of tensors are given, the pair must contain two tensors of shape
+            :math:`(N_{in}, D_{in_{src}})` and :math:`(N_{out}, D_{in_{dst}})`.
+
+            Note that in the special case of graph convolutional networks, if a pair of
+            tensors is given, the latter element will not participate in computation.
```
Comment (classicsong) [inline python/dgl/nn/mxnet/conv/graphconv.py:120]: Why there will be a pair of tensor?

File: python/dgl/sampling/neighbor.py [hunk 4]
```diff
@@ -166,4 +169,72 @@ def select_topk(g, k, weight, nodes=None, edge_dir='in', ascending=False):
...
+
+        If True, the edge IDs will be stored as ``dgl.EID`` feature for each edge type.
+
+    Examples
+    --------
+    To train a 3-layer GNN for node classification on a set of nodes ``train_nid`` on
+    a homogeneous graph where each node takes messages from all neighbors (assume
+    the backend is PyTorch):
+    >>> sampler = dgl.sampling.NeighborSampler([None, None, None])
+    >>> dataloader = dgl.sampling.NodeDataLoader(
```
Comment (classicsong) [inline python/dgl/sampling/neighbor.py:202]: Change doc string here.

File: python/dgl/transform.py [hunk 5]
```diff
@@ -987,6 +987,9 @@ def to_block(g, dst_nodes=None, include_dst_in_src=True):
             raise ValueError(
                 'Graph has more than one node type; please specify a dict for dst_nodes.')
         dst_nodes = {g.ntypes[0]: dst_nodes}
+    dst_nodes = {
+        ntype: utils.toindex(nodes, g._idtype_str).tousertensor()
```
Comment (classicsong) [inline python/dgl/transform.py:991]: Why we need to do this?

File: python/dgl/nn/mxnet/conv/graphconv.py (hunk 3 above)
Comment (BarclayII) [reply python/dgl/nn/mxnet/conv/graphconv.py:120]: Because one may still want to train GCN stochastically, which may feed the GCN module with a block (i.e. bipartite graph).

File: python/dgl/transform.py (hunk 5 above)
Comment (BarclayII) [reply python/dgl/transform.py:993]: Sometimes the arguments may be a list and we need to convert it to a tensor.

File: examples/pytorch/rgcn-hetero/entity_classify_mb.py (hunk 2 above)
Comment (BarclayII) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py:122]: Do you feel it's more intuitive than writing `seeds = ...`?  Should I also return the input nodes then?

Comment (classicsong) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py:122]: Yes, getting seed from loader is more intuitive to me

Comment (BarclayII) [reply examples/pytorch/rgcn-hetero/entity_classify_mb.py:122]: Then what about input nodes?

File: docs/source/api/python/sampling.rst (hunk 1 above)
Comment (BarclayII) [reply docs/source/api/python/sampling.rst:40]: Right.  Let me fix it...

**Comment by BarclayII** [comment]
Can we merge this?

File: python/dgl/sampling/neighbor.py [hunk 6]
```diff
@@ -166,4 +169,73 @@ def select_topk(g, k, weight, nodes=None, edge_dir='in', ascending=False):
...
+
+    Often works with a DGL dataloader for stochastic training of GNNs.
+
+    Parameters
+    ----------
+    fanouts : list[int] or list[dict[etype, int]]
+        List of neighbors to sample per edge type for each GNN layer, starting from the
+        first layer.
+
+        If the graph is homogeneous, only an integer is needed for each layer.
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/neighbor.py:186]: Updated.

File: python/dgl/sampling/neighbor.py [hunk 7]
```diff
@@ -166,4 +169,73 @@ def select_topk(g, k, weight, nodes=None, edge_dir='in', ascending=False):
...
+        Whether to return edge IDs of the original graph in the sampled blocks.
+
+        If True, the edge IDs will be stored as ``dgl.EID`` feature for each edge type.
+
+    Examples
+    --------
+    To train a 3-layer GNN for node classification on a set of nodes ``train_nid`` on
+    a homogeneous graph where each node takes messages from all neighbors (assume
+    the backend is PyTorch):
+    >>> sampler = dgl.sampling.NeighborSampler([None, None, None])
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/neighbor.py:199]: -1 is not supported in C interface.

[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/neighbor.py:199]: Just realized that we don't support -1 as all neighbors in C.

File: python/dgl/sampling/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+    for etype in block.canonical_etypes:
+        block.edges[etype].data[EID] = frontier.edges[etype].data[EID][
+            block.edges[etype].data[EID]]
+    return block
+
+class BlockSampler(object):
+    """Abstract class specifying the neighborhood sampling strategy for DGL data loaders.
+
+    The main method for BlockSampler is :func:`~dgl.sampling.BlockSampler.sample_blocks`,
+    which generates a list of blocks for a multi-layer GNN given a set of seed nodes to
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/dataloader.py:30]: It's in see also section

File: python/dgl/sampling/dataloader.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+    have their outputs computed.
+
+    The default implementation of :py:meth:`~dgl.sampling.BlockSampler.sample_blocks` is
+    to repeat ``__len__`` times the following:
+
+    * Obtain a frontier with the same nodes as the original graph but only the edges
+      involved in message passing on the last layer.
+      This is done with :py:meth:`~dgl.sampling.BlockSampler.sample_frontier`.
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/dataloader.py:38]: Not necessarily.  PinSAGE is a counterexample whose frontier is not a subgraph.

File: python/dgl/sampling/dataloader.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+
+        * Other arguments are the same ones passed into
+          :py:meth:`~dgl.sampling.BlockSampler.sample_blocks` method.
+
+        Parameters
+        ----------
+        postprocessor : callable
+            The postprocessor.
+        """
+        self._frontier_postprocessors.append(postprocessor)
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/dataloader.py:117]: One should be enough.  It shouldn't be abstract though as it should have a default implementation of doing nothing.

File: python/dgl/sampling/dataloader.py [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+      :py:meth:`~dgl.sampling.BlockSampler.add_block_postprocessor`.
+
+    * Prepend the block to the block list to be returned.
+
+    All subclasses should either
+
+    * Override :py:meth:`~dgl.sampling.BlockSampler.sample_blocks` method, or
+
+    * Override both :py:meth:`~dgl.sampling.BlockSampler.__len__` method and
+      :py:meth:`~dgl.sampling.BlockSampler.sample_frontier` method.
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/dataloader.py]: As we discussed before one can also choose to override `sample_blocks` (e.g. in case of KGCN and TGAT).

File: python/dgl/sampling/dataloader.py [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+        if isinstance(nids, Mapping):
+            self._dataset = utils.FlattenedDict(nids)
+        else:
+            self._dataset = nids
+
+    @property
+    def dataset(self):
+        return self._dataset
+
+    def collate(self, items):
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/dataloader.py:305]: Added docstring.

File: python/dgl/sampling/neighbor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sampling/neighbor.py:210]: Added the support.

**Comment by BarclayII** [comment]
Added code for minibatch inference on RGCN.  This is also related to #1695 .
